lib/evmone/eof.cpp

```diff
@@ -90,37 +151,273 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+/// Requires that the input is validated against truncation.
+bool validate_rjump_destinations(bytes_view code) noexcept
+{
+    // Collect relative jump destinations and immediate locations
+    const auto code_size = code.size();
+    // list of all possible absolute rjumps destinations positions
+    std::vector<size_t> rjumpdests;
+    // bool map of immediate arguments positions in the code
+    std::vector<bool> immediate_map(code_size);
+
+    const auto check_rjump_destination = [code_size](auto rel_offset_data_it,
+                                             size_t post_pos) -> std::optional<size_t> {
+        const auto rel_offset = read_int16_be(rel_offset_data_it);
+        const auto jumpdest = static_cast<int32_t>(post_pos) + rel_offset;
+        if (jumpdest < 0 || static_cast<size_t>(jumpdest) >= code_size)
+            return {};
+        else
+            return static_cast<size_t>(jumpdest);
+    };
+
+    for (size_t i = 0; i < code_size; ++i)
+    {
+        const auto op = code[i];
+        size_t imm_size = instr::traits[op].immediate_size;
+
+        if (op == OP_RJUMP || op == OP_RJUMPI)
+        {
+            if (auto jumpdest_opt = check_rjump_destination(&code[i + 1], i + 3); jumpdest_opt)
+                rjumpdests.push_back(jumpdest_opt.value());
+            else
+                return false;
+        }
+        else if (op == OP_RJUMPV)
+        {
+            constexpr auto REL_OFFSET_SIZE = sizeof(int16_t);
+
+            const auto count = code[i + 1];
+
+            imm_size += size_t{1} /* count */ + count * REL_OFFSET_SIZE /* tbl */;
+
+            const size_t post_pos = i + 1 + imm_size;
+
+            for (size_t k = 0; k < count * REL_OFFSET_SIZE; k += REL_OFFSET_SIZE)
+            {
+                auto jumpdest_opt =
+                    check_rjump_destination(&code[i + 1 + 1 + static_cast<uint16_t>(k)], post_pos);
+
+                if (jumpdest_opt)
+                    rjumpdests.push_back(jumpdest_opt.value());
+                else
+                    return false;
+            }
+        }
+
+        std::fill_n(immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, imm_size, true);
+        i += imm_size;
+    }
+
+    // Check relative jump destinations against immediate locations.
+    for (const auto rjumpdest : rjumpdests)
+        if (immediate_map[rjumpdest])
+            return false;
+
+    return true;
+}
+
+std::pair<EOFValidationError, int32_t> validate_max_stack_height(
+    bytes_view code, size_t func_index, const std::vector<EOFCodeType>& funcs_in_outs)
+{
+    assert(!code.empty());
+
+    // Stack height in the header is limited to uint16_t, but keeping larger size for ease of
+    // calculation.
+    std::vector<int32_t> stack_heights = std::vector<int32_t>(code.size(), -1);
+    std::stack<size_t> worklist;
+
+    int32_t stack_height = 0;
+    stack_heights[0] = funcs_in_outs[func_index].inputs;
+    worklist.push(0);
+
+    size_t i = 0;
+    Opcode opcode = {};
```

Comment:
`opcode`, `i` and `successors` can be moved into the loop, they are not used outside.